source/server/server.cc

```diff
@@ -310,11 +310,10 @@ void InstanceImpl::initialize(Options& options,
       runtime(), stats(), threadLocal(), random(), dnsResolver(), sslContextManager(), dispatcher(),
       localInfo(), secretManager(), api());
 
-  // Now the configuration gets parsed. The configuration may start setting thread local data
-  // per above. See MainImpl::initialize() for why we do this pointer dance.
-  Configuration::MainImpl* main_config = new Configuration::MainImpl();
-  config_.reset(main_config);
-  main_config->initialize(bootstrap_, *this, *cluster_manager_factory_);
+  // Now the configuration gets parsed. The configuration may start setting
+  // thread local data per above. See MainImpl::initialize() for why ConfigImpl
```

Comment:
I think the comment in configuraiton_impl.h at the declaration for initialize() is still accurate:
```
   * Initialize the configuration. This happens here vs. the constructor because the initialization
   * will call through the server into the config to get the cluster manager so the config object
   * must be created already.
```

In this PR the language about 'pointer dance' here in this file has been changed.

Reply:
Pedantically, I think it's a little different. Previously, since we were using a pointer, we needed to create the object then call initialize so the pointer is valid. Now, we create the config when we create the server, but we don't initialize it until we have the various pieces (factory) that we need. Personally I would alter to make it more clear but up to you.

Reply:
I gave it another shot; ptal. I'm not sure if I've adequately captured your concern.
```
   * MainImpl is created in two phases. In the first phase it is
   * default-constructed without a configuration as part of the server. The
   * server won't be fully populated yet. initialize() applies the
   * configuration in the second phase, as it requires a fully populated server.
```